File: backend/app/schemas/player.py
"""Player-area schemas: characters, campaign membership, notes."""

from datetime import datetime

from pydantic import BaseModel, Field

from app.schemas.character_rest import RestResult


class CharacterCreate(BaseModel):
    campaign_id: int
    name: str = Field(min_length=1, max_length=200)
    char_class: str = ""
    ancestry: str = ""
    community: str = ""
    level: int = Field(default=1, ge=1, le=20)
    extra: str = "{}"


class CharacterUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=200)
    char_class: str | None = None
    ancestry: str | None = None
    community: str | None = None
    level: int | None = Field(default=None, ge=1, le=20)
    extra: str | None = None

    model_config = {"extra": "forbid"}


class CharacterOut(BaseModel):
    id: int
    player_user_id: int
    campaign_id: int
    name: str
    char_class: str
    ancestry: str
    community: str
    level: int
    extra: str
    hp_marked: int
    stress_marked: int
    hope: int
    armor_slots_marked: int
    created_at: datetime


class MemberCampaignOut(BaseModel):
    id: int
    name: str
    description: str
    gm_user_id: int
    created_at: datetime
    active_session_room: str | None


class CampaignMemberOut(BaseModel):
    id: int
    campaign_id: int
    player_user_id: int
    player_username: str
    joined_at: datetime


class AddMemberRequest(BaseModel):
    username: str


class NoteOut(BaseModel):
    campaign_id: int
    body: str
    updated_at: datetime


class NoteUpdate(BaseModel):
    body: str = Field(max_length=20000)


class RestResponse(BaseModel):
    character: CharacterOut
    result: RestResult

File: frontend/src/api/player.ts
import { apiGet, apiPost, apiPut, apiPatch, apiDelete } from './client';

export interface MemberCampaign {
  id: number;
  name: string;
  description: string;
  gm_user_id: number;
  created_at: string;
}

export interface Character {
  id: number;
  player_user_id: number;
  campaign_id: number;
  name: string;
  char_class: string;
  ancestry: string;
  community: string;
  level: number;
  extra: string;
  hp_marked: number;
  stress_marked: number;
  hope: number;
  armor_slots_marked: number;
  created_at: string;
}

export interface CharacterState {
  hp_marked?: number;
  stress_marked?: number;
  hope?: number;
  armor_slots_marked?: number;
}

export interface Note {
  campaign_id: number;
  body: string;
  updated_at: string;
}

export type RestType = 'short' | 'long';
export type RestMove = 'tend_wounds' | 'clear_stress' | 'repair_armor' | 'prepare';

export interface RestResult {
  field: 'hp_marked' | 'stress_marked' | 'armor_slots_marked' | 'hope';
  roll: number | null;
  tier: number | null;
  amount: number;
  new_value: number;
}

export interface RestResponse {
  character: Character;
  result: RestResult;
}

export function listMyCampaigns(): Promise<MemberCampaign[]> {
  return apiGet('/api/player/campaigns');
}

export function listMyCharacters(campaignId?: number): Promise<Character[]> {
  const query = campaignId !== undefined ? `?campaign_id=${campaignId}` : '';
  return apiGet(`/api/player/characters${query}`);
}

export function createCharacter(input: {
  campaign_id: number;
  name: string;
  char_class: string;
  ancestry: string;
  community: string;
  level: number;
  extra?: string;
}): Promise<Character> {
  return apiPost('/api/player/characters', input);
}

export function updateCharacter(
  id: number,
  updates: Partial<
    Pick<Character, 'name' | 'char_class' | 'ancestry' | 'community' | 'level' | 'extra'>
  >,
): Promise<Character> {
  return apiPut(`/api/player/characters/${id}`, updates);
}

export function deleteCharacter(id: number): Promise<void> {
  return apiDelete(`/api/player/characters/${id}`);
}

export function updateCharacterState(id: number, state: CharacterState): Promise<Character> {
  return apiPatch(`/api/player/characters/${id}/state`, state);
}

export function restCharacter(
  id: number,
  rest_type: RestType,
  move: RestMove,
): Promise<RestResponse> {
  return apiPost(`/api/player/characters/${id}/rest`, { rest_type, move });
}

export function checkDowntimeAvailable(): Promise<{ available: boolean }> {
  return apiGet('/api/player/downtime');
}

export function getNote(campaignId: number): Promise<Note> {
  return apiGet(`/api/player/campaigns/${campaignId}/note`);
}

export function saveNote(campaignId: number, body: string): Promise<Note> {
  return apiPut(`/api/player/campaigns/${campaignId}/note`, { body });
}
